### PR TITLE
bluetooth: conn: Use a separate workqueue for connection TX notify

### PR DESCRIPTION
**Pull request by MarekPieta**
Use a separate workqueue instead of system workqueue for connection TX notify processing. This makes Bluetooth stack more independent from the system workqueue.

### CODE REVIEW

**Comment by Thalley** [comment]
This feels going in a different direction that a few other recent PRs where we've move BT things _to_ the system workqueue :o

**Comment by alwa-nordic** [comment]
The idea is fine as an option if you want it. But we are not going to spend time testing this upstream. I have not reviewed the code yet.

As an alternative, I propose we create a second system work queue, and explicitly state in its documentation that it must be treated isr-like and invoking any blocking functions is forbidden. @MarekPieta, does your application have RAM available for another system work queue?

**Comment by MarekPieta** [comment]
> This feels going in a different direction that a few other recent PRs where we've move BT things _to_ the system workqueue :o

It depends on what "things" you would like to do in the system workqueue context. From what I noticed so far, the Bluetooth stack may already perform settings (non-volatile memory) operations from the system workqueue context. Application also might perform some time consuming/settings-related operations from the system workqueue context. Such operations could have negative impact on the Bluetooth stack (delaying `tx_notify` processing).

Also there is a risk that synchronizing non-volatile memory and radio operations (required e.g. for nRF SoCs) could lead to problems if both Bluetooth and non-volatile memory operations are performed from the system workqueue context (and even Bluetooth stack performs settings operations from the system workqueue context). Blocking system workqueue during a non-volatile memory operation while waiting for Bluetooth to perform operation also in the system workqueue context could eventually even introduce a risk of deadlock related to flash synchronization mechanisms in some cases.

Because of that I think that using separate threads for Bluetooth may improve stability of the Bluetooth stack in some use-cases. Relying on system workqueue context for more complex use-cases may be problematic in general.

**Comment by MarekPieta** [comment]
> The idea is fine as an option if you want it. But we are not going to spend time testing this upstream. I have not reviewed the code yet.
> 
> As an alternative, I propose we create a second system work queue, and explicitly state in its documentation that it must be treated isr-like and invoking any blocking functions is forbidden. @MarekPieta, does your application have RAM available for another system work queue?

I think that introducing an additional work queue to separate operations that are more time critical may be step in the right direction (more complex applications could benefit from it; it could also be shared among subsystems). I think that my application could use it for most of the supported boards too (we should be able to fit the additional thread in memory).

Introducing the proposed system-wide solution is a bigger task and it would require more work. Could we consider using my proposed improvement until the system-wide solution is ready (using it as a temporary improvement)?

**Comment by MarekPieta** [comment]
Discussed offline with @alwa-nordic 

I will use a separate workqueue (available under a Kconfig option) to handle `tx_notify` as currently proposed solution also introduces risk of deadlocks in some edge cases. I will update the PR once I prepare the new solution.

File: subsys/bluetooth/host/Kconfig
Comment (rugeGerritsen) [inline subsys/bluetooth/host/Kconfig:137]: There should be tests verifying that the stack works with and without this option enabled, right?

Should there also be tests that verifies that ATT cannot time out/proves it does when the system workqueue is blocked for a long time?

Comment (alwa-nordic) [reply subsys/bluetooth/host/Kconfig:137]: I think we should just insist on having this enabled and remove the option. Doing so will remove the need for some hacks in the code like https://github.com/zephyrproject-rtos/zephyr/pull/79258/files#diff-00881f70fc7ec49b966fed79bd6a5a8ed5357599b241b5b029736bbc97683f83R275-R288

Comment (Thalley) [reply subsys/bluetooth/host/Kconfig:137]: Did we just remove one thread from BT and replace it with the system workqueue, and now adding a new thread to offload from the system workqueue?

Comment (MarekPieta) [reply subsys/bluetooth/host/Kconfig:137]: > I think we should just insist on having this enabled and remove the option. Doing so will remove the need for some hacks in the code like https://github.com/zephyrproject-rtos/zephyr/pull/79258/files#diff-00881f70fc7ec49b966fed79bd6a5a8ed5357599b241b5b029736bbc97683f83R275-R288

Removing the Kconfig option and making the feature enabled by default. I will also modify  https://github.com/zephyrproject-rtos/zephyr/pull/79258/files#diff-00881f70fc7ec49b966fed79bd6a5a8ed5357599b241b5b029736bbc97683f83R275-R288 (I noticed that you do not like this approach)

Comment (MarekPieta) [reply subsys/bluetooth/host/Kconfig:137]: > Did we just remove one thread from BT and replace it with the system workqueue, and now adding a new thread to offload from the system workqueue?

It seems that reusing system workqueue causes problems here. Because of that we may need to change approach here a bit.

Comment (alwa-nordic) [reply subsys/bluetooth/host/Kconfig:137]: > Did we just remove one thread from BT and replace it with the system workqueue, and now adding a new thread to offload from the system workqueue?

We went from removing a thread, turning it into a non-blocking work-item and placing it on the system work queue, to then having people noticing that blocking the system work queue actually blocks the system work queue. But since "it worked before", it's our fault and the conclusion is that we cannot rely on the system work queue to not be blocked. We are now adding a second system work queue that is explicitly only for non-blocking work.

Comment (alwa-nordic) [reply subsys/bluetooth/host/Kconfig:137]: > Right, but what we did gain from doing that then?

Architecturally, we removed a long-lived thread.

Ideally the new thread would not be in Bluetooth subsystem, but rather a service from the OS. There could be a kconfig "CONFIG_I_PROMISE_ONLY_SHORT_LIVED_TASKS_ON_SYSTEM_WORK_QUEUE". When selected the system work queue can be used.

> If we do need an internal workqueue for BT, then should we consider merging this with the BT_LONG_WQ? BT_LONG_WQ is not blocking, but may take a while to process, so not sure if we can use that

The long queue time will be a problem.

Comment (MarekPieta) [reply subsys/bluetooth/host/Kconfig]: > > Architecturally, we removed a long-lived thread and made it into a short-lived task.
> 
> Doesn't creating a new workqueue also create a long-living thread?
> 
> `k_work_queue_start` is basically just a wrapper around `k_thread_create` and `k_thread_start`, so we've just replaced a long-living thread with another long-living thread as far as I can tell.

I am worried that we may need to accept this approach (at least temporarily), because seemingly we cannot rely on the system work queue to not be blocked. Of course I agree that we could optimize the solution further later.

Comment (alwa-nordic) [reply subsys/bluetooth/host/Kconfig]: Yes. The the number of threads as a metric will tell you that nothing has changed.

But the new thread does not need to be exclusive to the work that "BT TX" did before. It can be shared with any other short-lived tasks. Potentially system wide.

Comment (Thalley) [reply subsys/bluetooth/host/Kconfig]: > But the new thread does not need to be exclusive to the work that "BT TX" did before. It can be shared with any other short-lived tasks. Potentially system-wide.

"The new thread" here referring to system-wide thread this PR isn't adding, right? 

What I am concerned about is that this is a temp patch (which it sounds like it is), fixing a problem that isn't referenced nor described, effectively going back to the BT stack still using 3 threads, where the thread/workqueue added by this PR is also exclusively for TX (as it is only used for the TX notify). 

We are thus once again modifying the thread where the TX callbacks are called from, which is IMO not something we should take lightly.

Comment (alwa-nordic) [reply subsys/bluetooth/host/Kconfig]: > We are thus once again modifying the thread where the TX callbacks are called from, which is IMO not something we should take lightly.

This is a valid point. I wasn't aware we are doing TX callbacks from tx_work. It looks like we are if the connection is disconnected or a bt_send error occurs. Thanks for the reminder. We should get rid of that somehow and consolidate the callback invocations on different thread. I maintain that blocking callbacks is stealing, and there is no way to fix this without green threads like in Go, but we can make the system slightly more resilient by having a "sacrificial" thread.

But switching over to a separate thread again will not make anything worse. Except, we might need to ensure the stack size is the same.

Comment (Thalley) [reply subsys/bluetooth/host/Kconfig]: > But switching over to a separate thread again will not make anything worse. Except, we might need to ensure the stack size is the same.

No, but what does it solve? 
The workqueue/thread added by this PR still calls the TX callback in `tx_notify_process`, so an application can still block that workqueue, so aren't we just back to the same? Sure, we reduce the number of places where the thread can be blocked by higher layers (as it's only a single callback), but it doesn't entirely prevent it.

**Comment by Thalley** [comment]
> Applied changes from offline discussion with @alwa-nordic

Just to be clear: It is still the case that when `tx_notify_process` is called in the context of `conn_tx_workq` then it still calls the TX callback which can still be blocked by a poorly implemented application, so I'm still unclear whether this is really fixing the issue if `conn_tx_workq` should never be blocked.

**Comment by MarekPieta** [comment]
> > Applied changes from offline discussion with @alwa-nordic
> 
> Just to be clear: It is still the case that when `tx_notify_process` is called in the context of `conn_tx_workq` then it still calls the TX callback which can still be blocked by a poorly implemented application, so I'm still unclear whether this is really fixing the issue if `conn_tx_workq` should never be blocked.

Using a separate workq at least mitigates negative impact of other works that rely on the system workqueue. Because of that, it might be useful for some of the applications. E.g. if the system workqueue is known to be used for tasks that may last long - like settings operations which could trigger NVM erases. The feature is optional, so if an application does not need it, it can simply keep the feature disabled.

The feature can also help with mitigating risks related to synchronizing non-volatile memory and radio operations for nRF SoCs (see https://github.com/zephyrproject-rtos/zephyr/pull/79258#issuecomment-2385749987 for details).

**Review by Thalley — CHANGES_REQUESTED** [review]
I understand the use case for this and the side effects. It's not a direction I'm happy about (basically re-adding a minimal TX thread again), but I have to assume that there's requirement that unfortunately isn't very clear here (no links to feature requests/issues). 

It seems like a temporary patch to a fundamental issue, and this solution only alleviates the issue and doesn't solve it. 

That being said, I won't be blocking this once my outstanding comments have been solved. 

Besides the comments, we should add an entry in the release notes stating that this was added, as the removal of the TX thread was added to the 3.7 release notes.

File: subsys/bluetooth/host/Kconfig
Comment (MarekPieta) [reply subsys/bluetooth/host/Kconfig:149]: I did it in the same way as for `CONFIG_BT_RX_PRIO` and `CONFIG_BT_DRIVER_RX_HIGH_PRIO`. Still, it would not hurt to make the priority user-configurable. Updated.

File: subsys/bluetooth/host/Kconfig
Comment (Thalley) [reply subsys/bluetooth/host/Kconfig:144]: I think a good case for testing a stack size would be to run one of the LE Audio sample and in the `bt_bap_stream_ops.sent` callback measure the stack usage. 

The reason why I think this is a good way of measuring it, is that the callback goes from the work item in `conn.c` to the `sent` callback in `iso.c` to the `sent` callback in BAP and then finally to the application (possible passing through the `sent` callback in `cap_stream.c` as well.

File: subsys/bluetooth/host/Kconfig
Comment (Thalley) [reply subsys/bluetooth/host/Kconfig:149]: Yeah, checked those out too, and didn't really understand why they were non-configurable Kconfig options either :)

**Comment by LingaoM** [comment]
Why not borrow BT RX? This is also a workq, and the stack size is large enough.

File: subsys/bluetooth/host/Kconfig
Comment (alwa-nordic) [reply subsys/bluetooth/host/Kconfig:144]: At the Bluetooth WG meeting today, there was consensus to default this setting to copy the value from the system work queue.

File: subsys/bluetooth/host/Kconfig
Comment (alwa-nordic) [inline subsys/bluetooth/host/Kconfig:132]: The PR mostly looks ok now, so I'll do a nitpick: I would like to rename this to `BT_CONN_TX_NOTIFY_WQ` to avoid confusing it with `tx_work`.

File: subsys/bluetooth/host/Kconfig
Comment (MarekPieta) [reply subsys/bluetooth/host/Kconfig:144]: Ok, updated.

File: subsys/bluetooth/host/Kconfig
Comment (MarekPieta) [reply subsys/bluetooth/host/Kconfig:149]: I assumed that it's done to prevent the user from breaking something while still allowing to override the value by redefining the default in application's Kconfig (that's why I initially followed the same approach here)

File: subsys/bluetooth/host/conn.c
Comment (MarekPieta) [reply subsys/bluetooth/host/conn.c:4181]: You are right. I made it static to avoid allocating the structure on stack and thanks to that save a few bytes of FLASH.
Anyway, changed as suggested.

File: subsys/bluetooth/host/Kconfig
Comment (MarekPieta) [reply subsys/bluetooth/host/Kconfig:132]: Changed

**Comment by MarekPieta** [comment]
> Why not borrow BT RX? This is also a workq, and the stack size is large enough.

Unfortunately that might not work properly too. I already discussed this approach with @alwa-nordic. He explained to me that it might cause deadlocks because the work posted to BT WQ may be blocking - e.g. it invokes application callbacks, ATT server (which blocks on allocation), etc.
Because of that, the idea was eventually rejected.

File: subsys/bluetooth/host/Kconfig
Comment (Thalley) [reply subsys/bluetooth/host/Kconfig]: That pretty much sounds like "We want it to be configurable but not really" :D Or "Configurable but with extra steps". 

`int` Kconfigs are generally just a pain to work with, as they are not as flexible as `bool`s (`imply` `select` only works with `bool`s).

File: doc/releases/release-notes-4.0.rst
Comment (Thalley) [inline doc/releases/release-notes-4.0.rst:111]: Do we have max line length requirements for .rst files?

File: subsys/bluetooth/host/Kconfig
Comment (Thalley) [inline subsys/bluetooth/host/Kconfig:134]:
```suggestion
config BT_CONN_TX_NOTIFY_WQ
	bool "Use a separate workqueue for connection TX notify processing [EXPERIMENTAL]"
	depends on BT_CONN_TX
	select EXPERIMENTAL
```

**Review by Thalley — DISMISSED** [review]
No objects from me anymore, but as discussed in the BT WG, this is a half-asses solution to a problem caused by an application's usage of the system workqueue. 

There are probably better solutions to this, but less trivial to implement than this. 

I am OK with adding this for now, but let's mark it as experimental so we can easily remove it again

File: doc/releases/release-notes-4.0.rst
Comment (MarekPieta) [reply doc/releases/release-notes-4.0.rst:111]: I see that generally doc lines tend to be split if they are longer than 100 characters (but it's not always a case). I will align my note.

Comment (Thalley) [reply doc/releases/release-notes-4.0.rst:111]: @kartben Thoughts?

File: subsys/bluetooth/host/Kconfig
Comment (rugeGerritsen) [reply subsys/bluetooth/host/Kconfig]: Did you decide upon that you didn't want to add any form of test coverage as part of this PR?

Comment (Thalley) [reply subsys/bluetooth/host/Kconfig]: @rugeGerritsen I think we should definitely have some tests in BSIM using this.

Comment (MarekPieta) [reply subsys/bluetooth/host/Kconfig]: If needed, maybe I could open a separate PR that enables the feature by default to run all of the CI tests with the feature enabled (for additional validation of the feature)?

**Review by alwa-nordic — APPROVED** [review]

File: subsys/bluetooth/host/Kconfig
Comment (Thalley) [reply subsys/bluetooth/host/Kconfig]: I would argue that we should run at least _some_ tests in this PR, otherwise we are adding a untested Kconfig. It makes more sense to test before you merge a feature, rather than after :D

Comment (MarekPieta) [reply subsys/bluetooth/host/Kconfig]: Let's enable feature by default on a separate test PR to run CI tests on it then: https://github.com/zephyrproject-rtos/zephyr/pull/79713

Comment (Thalley) [reply subsys/bluetooth/host/Kconfig]: > Let's enable feature by default on a separate test PR to run CI tests on it then: #79713

Can you explain why you are not adding tests in this PR? We need the tests regardless

File: doc/releases/release-notes-4.0.rst
Comment (kartben) [reply doc/releases/release-notes-4.0.rst]: Ya in theory it's 100. In practice a better rule would probably be one sentence per line, as it also makes it easier for folks to realize that a blank line is required for paragraph breaks -- when using line wrapping people tend to assume that having just one line return will do that (as it looks "ok enough" in source form) when obviously it doesn't. It also makes diffs cleaner too, since adding a word or two to a line really just impacts that line as opposed to potentially many if it "pushes" a wrapped line to the next (and so on).
Unfortunately switching to this approach would require updating (and potentially breaking) so many files that I have yet to give it a try :)

In the meantime, 100 characters per line it is, when possible :)

**Review by Thalley — APPROVED** [review]
Approving on the condition that @MarekPieta or someone else involved will add proper tests of this soon after.

**Review by rugeGerritsen — CHANGES_REQUESTED** [review]
Can you update the commit message with the following:
* The motivation behind the change
  * What use case will start working when enabling this config/stopped working when the TX queue was removed
  * Why the configuration is not enabled by default
  * Why it is experimental
 
Document the plans going forward:
* How and who is going to test this?
* Are we planning to get rid of it afterwards?

That will make it easier to look at this change again some time later

**Comment by Thalley** [comment]
@MarekPieta @alwa-nordic 
Arguably we should discuss this further in a GH discussion or on Discord, but just to "answer" the above: Has the idea of being able to configure multiple (identical) system workqueues with a load balancer been considered? It sounds like overkill in an embedded system, but could in theory have fixed this issue as well, and could work as a scalable solution compared to the suggested non-blocking workqueue (how would we even enforce that?).

**Comment by jhedberg** [comment]
> Has the idea of being able to configure multiple (identical) system workqueues with a load balancer been considered?

That's an interesting idea, though in some ways it sounds like it's intruding on the responsibility area of the scheduler. I wouldn't limit it to "the system", rather make it a general "workqueue pool" style API, and then the system work queue could potentially be extended to take advantage of it. Instead of having a 1:1 mapping between a queue and a thread that runs the work, you'd have a 1:N mapping from a queue to multiple possible threads that can execute the work. Looking at `kernel/work.c` it seems like it might not even be that hard to implement. Most likely it's not quite as simple as starting multiple `work_queue_main()` threads for the same queue, however it's possible that not much more is needed. Anyway, if you want to drive this, it's something that should be proposed on a Zephyr architecture level, i.e. create an issue for it and bring it to the Architecture WG for discussion.

**Pull request merged**
